src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1130,47 +1130,41 @@ private IIfStatement CreateBoundIfStatementOperation(BoundIfStatement boundIfSta
         }
 
-        private IWhileUntilLoopStatement CreateBoundWhileStatementOperation(BoundWhileStatement boundWhileStatement)
+        private IWhileLoopStatement CreateBoundWhileStatementOperation(BoundWhileStatement boundWhileStatement)
         {
-            bool isTopTest = true;
-            bool isWhile = true;
             Lazy<IOperation> condition = new Lazy<IOperation>(() => Create(boundWhileStatement.Condition));
-            LoopKind loopKind = LoopKind.WhileUntil;
             Lazy<IOperation> body = new Lazy<IOperation>(() => Create(boundWhileStatement.Body));
             SyntaxNode syntax = boundWhileStatement.Syntax;
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
             bool isImplicit = boundWhileStatement.WasCompilerGenerated;
-            return new LazyWhileUntilLoopStatement(isTopTest, isWhile, condition, loopKind, body, _semanticModel, syntax, type, constantValue, isImplicit);
+            return new LazyWhileLoopStatement(condition, body, _semanticModel, syntax, type, constantValue, isImplicit);
         }
 
-        private IWhileUntilLoopStatement CreateBoundDoStatementOperation(BoundDoStatement boundDoStatement)
+        private IDoLoopStatement CreateBoundDoStatementOperation(BoundDoStatement boundDoStatement)
         {
-            bool isTopTest = false;
-            bool isWhile = true;
+            DoLoopKind doLoopKind = DoLoopKind.DoWhileBottomLoop;
             Lazy<IOperation> condition = new Lazy<IOperation>(() => Create(boundDoStatement.Condition));
-            LoopKind loopKind = LoopKind.WhileUntil;
             Lazy<IOperation> body = new Lazy<IOperation>(() => Create(boundDoStatement.Body));
             SyntaxNode syntax = boundDoStatement.Syntax;
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
             bool isImplicit = boundDoStatement.WasCompilerGenerated;
-            return new LazyWhileUntilLoopStatement(isTopTest, isWhile, condition, loopKind, body, _semanticModel, syntax, type, constantValue, isImplicit);
+            return new LazyDoLoopStatement(doLoopKind, condition, body, _semanticModel, syntax, type, constantValue, isImplicit);
         }
 
         private IForLoopStatement CreateBoundForStatementOperation(BoundForStatement boundForStatement)
         {
             Lazy<ImmutableArray<IOperation>> before = new Lazy<ImmutableArray<IOperation>>(() => ToStatements(boundForStatement.Initializer));
+            Lazy<IOperation> condition = new Lazy<IOperation>(() => Create(boundForStatement.Condition));
             Lazy<ImmutableArray<IOperation>> atLoopBottom = new Lazy<ImmutableArray<IOperation>>(() => ToStatements(boundForStatement.Increment));
             ImmutableArray<ILocalSymbol> locals = boundForStatement.OuterLocals.As<ILocalSymbol>();
```

Comment:
>OuterLocals [](start = 68, length = 11)

How do we expose InnerLocals?

Reply:
Filed [https://github.com/dotnet/roslyn/issues/21823](https://github.com/dotnet/roslyn/issues/21823)

---
In reply to: [135970169](https://github.com/dotnet/roslyn/pull/21791#discussion_r135970169) [](ancestors = 135970169)